debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java

```diff
@@ -0,0 +1,358 @@
+            queue.offer(event);
+        }
+    }
+
+    public static <TResult> BufferingChangeStreamCursor<TResult> fromIterable(
+                                                                              ChangeStreamIterable<TResult> stream,
+                                                                              MongoDbTaskContext taskContext,
+                                                                              MongoDbStreamingChangeEventSourceMetrics metrics,
+                                                                              Clock clock) {
+        var config = taskContext.getConnectorConfig();
+
+        return new BufferingChangeStreamCursor<>(
+                new EventFetcher<>(stream, config.getMaxBatchSize(), metrics, clock, config.getPollInterval()),
+                Threads.newFixedThreadPool(MongoDbConnector.class, taskContext.serverName(), "replicator-buffer", 1),
+                config.getPollInterval());
+    }
+
+    /**
+     * Creates new prefetching cursor
+     *
+     * @param fetcher MongoDB change event fetcher
+     * @param executor executor used to dispatch buffering thread
+     * @param throttler throttling mechanism
+     */
+    public BufferingChangeStreamCursor(EventFetcher<TResult> fetcher, ExecutorService executor, DelayStrategy throttler) {
+        this.fetcher = fetcher;
+        this.executor = executor;
+        this.throttler = throttler;
+    }
+
+    public BufferingChangeStreamCursor(EventFetcher<TResult> fetcher, ExecutorService executor, Duration throttleMaxSleep) {
+        this(fetcher, executor, DelayStrategy.boundedExponential(Duration.ofMillis(1), throttleMaxSleep, 2));
+    }
+
+    public BufferingChangeStreamCursor<TResult> start() {
+        executor.submit(fetcher);
+        return this;
+    }
+
+    @Override
+    public ResumableChangeStreamEvent<TResult> tryNext() {
+        var event = pollWithDelay();
+        if (event != null) {
+            lastResumeToken = event.resumeToken;
+        }
+        return event;
+    }
+
+    /**
+     * Returns next event in buffer.
+     * Not that unlike other Mongo implementation this method does not block
+     *
+     * @throws NoSuchElementException when no element is available
+     * @return event
+     */
+    @Override
+    public ResumableChangeStreamEvent<TResult> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+        return tryNext();
+    }
+
+    /**
+     * Repeatedly polling fetcher for new event, exponentially waiting between polls until limit is reached
+     *
+     * @return event or null if not available within time limit
+     */
+    private ResumableChangeStreamEvent<TResult> pollWithDelay() {
+        boolean slept;
+        ResumableChangeStreamEvent<TResult> event;
+
+        do {
+            event = fetcher.poll();
+            slept = throttler.sleepWhen(event == null);
+        } while (slept);
+
+        return event;
+    }
```

Comment:
@jpechane I asked  @Naros to also take a a look and one thing he is worried about is blocking in the main thread  as it can block connector shutdown for up to poll interval. 

What's your take on this? Maybe the queue wouldn't mind being poll constantly and we should indeed just use some microsleep intervals.

Reply:
I wouldn't deviate from what you have if it works. The hard sleep in the main loop based on the poll interval and its impact on shutdown, when a user has a considerable poll time, is a rare use case and one I'm fine tackling as a separate concern as it would affect all our connectors anyway.

Reply:
@jcechace Is not this interruptible? If yes than it can either end up in `InterruptedException` or to be even more sure the loop can have additional condition relatied to `isInterrupted` or a running flag.

Reply:
@jpechane I thought it was ... @Naros  seems to think that the DelayStrategy when sleeping will postpone connector shutdown.